feat: integrate multichain assets rates controller to extension UI

salimtb opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**
This PR aims to integrate the new multichainAssetsRatesControllers into the extension. To achieve this, an upgrade of the assets controller to version 49 is required. As a result, the target branch is set accordingly:
[MetaMask PR #30250](https://github.com/MetaMask/metamask-extension/pull/30250).

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/30291?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. open a flask build and add a solana account
2. check if the polling for the rates is running

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

sahar-fehri committed:
fix: bump assets-controllers to v49

metamaskbot committed:
Update LavaMoat policies

sahar-fehri committed:
fix: dedupe

metamaskbot committed:
Update LavaMoat policies

sahar-fehri committed:
fix: merge conflicts

sahar-fehri committed:
fix: bump assets

sahar-fehri committed:
fix: patch

sahar-fehri committed:
fix: fix yarn

sahar-fehri committed:
fix: fix test

metamaskbot committed:
Update LavaMoat policies

sahar-fehri committed:
fix: fix patch

metamaskbot committed:
Update LavaMoat policies

sahar-fehri committed:
Merge branch 'main' into feat/bump-assets-controller-49.0.0

zone-live committed:
Merge branch 'main' into feat/bump-assets-controller-49.0.0

GuillaumeRx commented on app/scripts/metamask-controller.js:
```diff
@@ -1075,6 +1075,32 @@ export default class MetamaskController extends EventEmitter {
       disabled: !this.preferencesController.state.useCurrencyRateCheck,
     });
 
+    ///: BEGIN:ONLY_INCLUDE_IF(build-flask)
+    const multichainAssetsRatesControllerMessenger =
+      this.controllerMessenger.getRestricted({
+        name: 'MultiChainAssetsRatesController',
+        allowedActions: [
+          'AccountsController:listMultichainAccounts',
+          'SnapController:handleRequest',
+          'CurrencyRateController:getState',
+          'MultichainAssetsController:getState',
+          'MultichainBalancesController:getState',
+        ],
+        allowedEvents: [
+          'AccountsController:accountAdded',
+          'KeyringController:lock',
+          'KeyringController:unlock',
+          'CurrencyRateController:stateChange',
+          'MultichainAssetsController:stateChange',
+        ],
+      });
+
+    this.multichainAssetsRatesController = new MultiChainAssetsRatesController({
+      messenger: multichainAssetsRatesControllerMessenger,
+      state: initState.MultiChainAssetsRatesController,
+    });
+    ///: END:ONLY_INCLUDE_IF
+
```
Please use the new controller initialization logic :pray:

sahar-fehri committed:
Merge branch 'main' into feat/bump-assets-controller-49.0.0

sahar-fehri committed:
fix: lint

GuillaumeRx committed:
[WIP] `MultichainAssetsController` init

GuillaumeRx committed:
Update multichain controllers init
- Add `MultichainAssetsController` to the controller init list
- Fix issues with `MultichainBalanceController`
- Fix issues with `MultichainTransactionsController`
- Revert type changes

salimtb committed:
fix: integrate multichainAssetsRatesController

sahar-fehri committed:
Merge branch 'main' into feat/bump-assets-controller-49.0.0

salimtb committed:
fix: fix after rebase

sahar-fehri committed:
Merge branch 'main' into feat/bump-assets-controller-49.0.0

salimtb committed:
fix: fix linter

GuillaumeRx commented on app/scripts/controller-init/messengers/multichain/multichain-assets-rates-controller-messenger.ts:
```diff
@@ -0,0 +1,61 @@
+>;
+
+/**
+ * Get a restricted messenger for the Multichain Assets controller. This is scoped to the
```
```suggestion
 * Get a restricted messenger for the Multichain Assets Rate controller. This is scoped to the
```

GuillaumeRx commented on app/scripts/controller-init/messengers/multichain/multichain-assets-rates-controller-messenger.ts:
```diff
@@ -0,0 +1,61 @@
+
+/**
+ * Get a restricted messenger for the Multichain Assets controller. This is scoped to the
+ * actions and events that the multichain Assets controller is allowed to handle.
```
```suggestion
 * actions and events that the multichain Assets Rate controller is allowed to handle.
```

GuillaumeRx commented on app/scripts/controller-init/multichain/multichain-rates-assets-controller-init.ts:
```diff
@@ -0,0 +1,25 @@
+import { MultichainAssetsRatesControllerMessenger } from '../messengers/multichain';
+
+/**
+ * Initialize the Multichain Assets controller.
```
```suggestion
 * Initialize the Multichain Assets Rate controller.
```

GuillaumeRx commented on ui/selectors/multichain-assets-rates.ts:
```diff
@@ -0,0 +1,15 @@
+};
+
+/**
+ * Gets non-EVM accounts assets.
```
```suggestion
 * Gets non-EVM accounts assets rates.
```

salimtb committed:
fix: fix PR comments

GuillaumeRx commented:
Looking good :100:

GuillaumeRx reviewed (dismissed):
You need to omit this state key from sentry

salimtb committed:
Merge branch 'main' into salim/integrate-mutichain-assets-rates-controller

salimtb committed:
fix: fix e2e tests

salimtb committed:
fix: fix build

salimtb committed:
fix: fix build

salimtb committed:
fix: fix linter

sahar-fehri commented on ui/selectors/multichain-assets-rates.ts:
```diff
@@ -0,0 +1,15 @@
+ * @returns An object containing non-EVM assets per accounts.
+ */
+export function getAssetsRates(state: AssetsState) {
+  return state.metamask.conversionRates;
```
Do you think we should move these selectors to ui/selectors/assets.ts instead?

salimtb replied:
I prefer to keep these selectors separate because additional functions, such as retrieving metadata, will be added in the future. Keeping them isolated ensures better maintainability and testability. If we move them to `ui/selectors/assets.ts`, we would also need to relocate the tests, which could reduce modularity and make testing more complex.

salimtb replied:
from a code quality perspective, maintaining a clear separation of concerns helps with readability, maintainability, and scalability.

GuillaumeRx commented on app/scripts/controller-init/messengers/index.ts:
```diff
@@ -39,6 +40,9 @@ export const CONTROLLER_MESSENGERS = {
   MultichainTransactionsController: {
     getMessenger: getMultichainTransactionsControllerMessenger,
   },
+  MultiChainAssetsRatesController: {
+    getMessenger: getMultiChainAssetsRatesControllerMessenger,
+  },
```
```suggestion
  MultiChainAssetsRatesController: {
    getMessenger: getMultiChainAssetsRatesControllerMessenger,
    getInitMessenger: noop,
  },
```

GuillaumeRx commented on app/scripts/controller-init/multichain/multichain-rates-assets-controller-init.test.ts:
```diff
@@ -0,0 +1,51 @@
+> {
+  const baseControllerMessenger = new Messenger();
+
+  return {
+    ...buildControllerInitRequestMock(),
+    controllerMessenger: getMultiChainAssetsRatesControllerMessenger(
+      baseControllerMessenger,
+    ),
+  };
```
```suggestion
  return {
    ...buildControllerInitRequestMock(),
    controllerMessenger: getMultiChainAssetsRatesControllerMessenger(
      baseControllerMessenger,
    ),
    initMessenger: undefined
  };
```

GuillaumeRx reviewed (changes requested):
You need to add the controller name here: https://github.com/MetaMask/metamask-extension/blob/main/app/scripts/controller-init/utils.ts#L39

You might want to wait for https://github.com/MetaMask/metamask-extension/pull/30301to land on main as it will create a merge conflict

sahar-fehri commented:
Can re-approve once conflicts are resolved after https://github.com/MetaMask/metamask-extension/pull/30301

salimtb committed:
fix: fix PR comments

salimtb committed:
Merge branch 'main' into salim/integrate-mutichain-assets-rates-controller

salimtb committed:
fix: fix linter

GuillaumeRx reviewed (changes requested):
Still missing the controller's name here: https://github.com/MetaMask/metamask-extension/blob/main/app/scripts/controller-init/utils.ts#L39

salimtb committed:
fix: add controller name

salimtb commented:
> You need to add the controller name here: https://github.com/MetaMask/metamask-extension/blob/main/app/scripts/controller-init/utils.ts#L39
> 
> You might want to wait for https://github.com/MetaMask/metamask-extension/pull/30301to land on main as it will create a merge conflict

resolved

GuillaumeRx commented on app/scripts/controller-init/messengers/index.ts:
```diff
@@ -45,6 +46,10 @@ export const CONTROLLER_MESSENGERS = {
     getMessenger: getMultichainTransactionsControllerMessenger,
     getInitMessenger: noop,
   },
+  MultiChainAssetsRatesController: {
+    getMessenger: getMultiChainAssetsRatesControllerMessenger,
+    getInitMessenger: noop,
+  },
```
Please sort alphabetically

GuillaumeRx commented on app/scripts/controller-init/messengers/multichain/index.ts:
```diff
@@ -1,7 +1,9 @@
 export { getMultichainAssetsControllerMessenger } from './multichain-assets-controller-messenger';
 export { getMultichainBalancesControllerMessenger } from './multichain-balances-controller-messenger';
 export { getMultichainTransactionsControllerMessenger } from './multichain-transactions-controller-messenger';
+export { getMultiChainAssetsRatesControllerMessenger } from './multichain-assets-rates-controller-messenger';
 
 export type { MultichainAssetsControllerMessenger } from './multichain-assets-controller-messenger';
 export type { MultichainBalancesControllerMessenger } from './multichain-balances-controller-messenger';
 export type { MultichainTransactionsControllerMessenger } from './multichain-transactions-controller-messenger';
+export type { MultiChainAssetsRatesControllerMessenger } from './multichain-assets-rates-controller-messenger';
```
Same here

GuillaumeRx commented on app/scripts/metamask-controller.js:
```diff
@@ -2056,6 +2057,7 @@ export default class MetamaskController extends EventEmitter {
       MultichainAssetsController: MultichainAssetsControllerInit,
       MultichainBalancesController: MultichainBalancesControllerInit,
       MultichainTransactionsController: MultichainTransactionsControllerInit,
+      MultiChainAssetsRatesController: MultiChainAssetsRatesControllerInit,
```
same here

GuillaumeRx commented on app/scripts/controller-init/controller-list.ts:
```diff
@@ -75,6 +77,7 @@ export type ControllerFlatState = AccountsController['state'] &
   MultichainAssetsController['state'] &
   MultichainBalancesController['state'] &
   MultichainTransactionsController['state'] &
+  MultiChainAssetsRatesController['state'] &
```
Same here

GuillaumeRx commented on app/scripts/controller-init/controller-list.ts:
```diff
@@ -44,6 +45,7 @@ export type Controller =
   | MultichainAssetsController
   | MultichainBalancesController
   | MultichainTransactionsController
+  | MultiChainAssetsRatesController
```
same here

salimtb committed:
fix: sort alphabetically the added controller

salimtb committed:
fix: sort initialisation

zone-live commented on ui/selectors/assets.ts:
```diff
@@ -23,3 +30,13 @@ export function getAccountAssets(state: AssetsState) {
+ * @param state - Redux state object.
+ * @returns An object containing non-EVM assets per accounts.
+ */
+export function getAssetsRates(state: AssetsRatesState) {
```
Quick question, will this selector replace the existing [getMultichainCoinRates](https://github.com/MetaMask/metamask-extension/blob/main/ui/selectors/multichain.ts#L400)

salimtb replied:
@zone-live the one you mentioned is only for Native tokens , while the one on this PR is for SPL tokens.
so the answer is no but we still to use both on the UI

salimtb committed:
Merge branch 'main' into salim/integrate-mutichain-assets-rates-controller
